### PR TITLE
Activity failure

### PR DESCRIPTION
**Pull request by vitarb**
This change adds support for activity failure into the core sdk.
Tested in 3 variations - integ test, lib.rs top level test and state machine test.

Sample history for non-retryable failure case:
```
❯ tctl wf show --wid 1012869045

   1  WorkflowExecutionStarted    {WorkflowType:{Name:test-workflow},                           
                                  ParentInitiatedEventId:0,                                     
                                  TaskQueue:{Name:activity_failed_workflow_1964480726,          
                                  Kind:Normal}, WorkflowTaskTimeout:10s,                        
                                  Initiator:Unspecified,                                        
                                  OriginalExecutionRunId:815b9c1e-0893-491f-b35f-20b47b470c72,  
                                  FirstExecutionRunId:815b9c1e-0893-491f-b35f-20b47b470c72,     
                                  Attempt:1, FirstWorkflowTaskBackoff:0s}                       
   2  WorkflowTaskScheduled       {TaskQueue:{Name:activity_failed_workflow_1964480726,         
                                  Kind:Normal}, StartToCloseTimeout:10s, Attempt:1}             
   3  WorkflowTaskStarted         {ScheduledEventId:2, Identity:integ_tester,                   
                                  RequestId:336090d5-6cc6-4745-b25c-c1ffc2a1a382}               
   4  WorkflowTaskCompleted       {ScheduledEventId:2,                                          
                                  StartedEventId:3,                                             
                                  Identity:integ_tester}                                        
   5  ActivityTaskScheduled       {ActivityId:3202999672,                                       
                                  ActivityType:{Name:test_activity},                            
                                  Namespace:default,                                            
                                  TaskQueue:{Name:activity_failed_workflow_1964480726,          
                                  Kind:Normal}, ScheduleToCloseTimeout:1m0s,                    
                                  ScheduleToStartTimeout:30s,                                   
                                  StartToCloseTimeout:30s, HeartbeatTimeout:30s,                
                                  WorkflowTaskCompletedEventId:4,                               
                                  RetryPolicy:{InitialInterval:1s,                              
                                  BackoffCoefficient:2, MaximumInterval:1m40s,                  
                                  MaximumAttempts:0}}                                           
   6  ActivityTaskStarted         {ScheduledEventId:5, Identity:integ_tester,                   
                                  RequestId:af393dfa-86cb-493a-a5ba-bc688d2c2841,               
                                  Attempt:1}                                                    
   7  ActivityTaskFailed          {Failure:{Message:activity failed,                            
                                  FailureInfo:{ApplicationFailureInfo:{NonRetryable:true}}},    
                                  ScheduledEventId:5, StartedEventId:6,                         
                                  Identity:integ_tester, RetryState:NonRetryableFailure}        
   8  WorkflowTaskScheduled       {TaskQueue:{Name:activity_failed_workflow_1964480726,         
                                  Kind:Normal}, StartToCloseTimeout:10s, Attempt:1}             
   9  WorkflowTaskStarted         {ScheduledEventId:8, Identity:integ_tester,                   
                                  RequestId:c046f40c-dc29-4b57-8d31-f8f4737a2a3a}               
  10  WorkflowTaskCompleted       {ScheduledEventId:8,                                          
                                  StartedEventId:9,                                             
                                  Identity:integ_tester}                                        
  11  WorkflowExecutionCompleted  {WorkflowTaskCompletedEventId:10} 
```

### CODE REVIEW

File: src/machines/activity_state_machine.rs [hunk 1]
```diff
@@ -192,6 +214,19 @@ impl WFMachinesAdapter for ActivityMachine {
                 }),
             }
             .into()],
+            ActivityMachineCommand::Fail(failure) => vec![ResolveActivity {
+                activity_id: self.shared_state.attrs.activity_id.clone(),
+                result: Some(ActivityResult {
+                    status: Some(activity_result::Status::Failed(ar::Failure {
+                        failure: match failure {
+                            Some(f) => Some(f.into()),
+                            None => None,
```
[the remark this answers is not in the document]
Comment (vitarb) [reply src/machines/activity_state_machine.rs:223]: It is, but let's check with Spencer, maybe there is a nicer way to write it.

[the remark this answers is not in the document]
Comment (vitarb) [reply src/machines/activity_state_machine.rs:223]: Turns out we can do this instead:
                        failure: failure.map(Into::into),

File: src/machines/activity_state_machine.rs [hunk 2]
```diff
@@ -149,7 +160,18 @@ impl TryFrom<HistoryEvent> for ActivityMachineEvents {
...
                 }
             }
-            Some(EventType::ActivityTaskFailed) => Self::ActivityTaskFailed,
+            Some(EventType::ActivityTaskFailed) => {
+                if let Some(history_event::Attributes::ActivityTaskFailedEventAttributes(attrs)) =
+                    e.attributes
+                {
+                    Self::ActivityTaskFailed(attrs)
+                } else {
+                    return Err(WFMachinesError::MalformedEvent(
```
[the remark this answers is not in the document]
Comment (vitarb) [reply src/machines/activity_state_machine.rs:169]: It's on the history event replay path though, so workflow error does make more sense. Also it's consistent with what we do for the timer, which is a pretty similar use case.

File: src/machines/activity_state_machine.rs [hunk 3]
```diff
@@ -337,9 +378,15 @@ impl StartedActivityCancelEventRecorded {
             Completed::default(),
         )
     }
-    pub(super) fn on_activity_task_failed(self) -> ActivityMachineTransition {
+    pub(super) fn on_activity_task_failed(
+        self,
+        attrs: ActivityTaskFailedEventAttributes,
+    ) -> ActivityMachineTransition {
         // notify_failed
```
Comment (Sushisource) [inline src/machines/activity_state_machine.rs:382]: We can probably clean up a bunch of these old comments

**Review by Sushisource — APPROVED** [review]

File: src/machines/activity_state_machine.rs (hunk 3 above)
Comment (vitarb) [reply src/machines/activity_state_machine.rs:382]: I actually like them as they map it directly to java handlers.

File: src/machines/activity_state_machine.rs [hunk 4]
```diff
@@ -1,4 +1,8 @@
+#![allow(clippy::large_enum_variant)]
```
[the remark this answers is not in the document]
Comment (vitarb) [reply src/machines/activity_state_machine.rs:1]: It's in the state machine, I don't think you can apply it there, same applies to the timer where you've added it on top too.

[the remark this answers is not in the document]
Comment (Sushisource) [reply src/machines/activity_state_machine.rs:1]: You can apply it just on top of the macro I believe

**Pull request merged**